packages/vite/src/node/baseEnvironment.ts

```diff
@@ -64,6 +65,7 @@ export class PartialEnvironment {
     this.name = name
     this._topLevelConfig = topLevelConfig
     this._options = options
+    const packageCache: PackageCache = new Map()
```

Comment:
I think this is going to far. The PackageCache has two mixed caches:
a. The path -> `PackageData` cache
b. the resolved cache in each `PackageData` (the `getResolvedCache`/`setResolvedCache` pair)

We still need (a) to be per-config and not per-environment. This cache doesn't depend on environment-specific options. We need to make only (b) per-environment.

What I was thinking is that `PackageCache` will remain as-is, the cache from (b) is removed from `PackageData`. And a new `ResolvedCache` is created per-environment that will expose `resolvedCache.get(packageData, key)`. This cache is passed to the resolve plugin that is the only place where it is used.

Reply:
IIUC making (b) per-environment is not enough. It also needs to be separated whether it's a resolve for `require` or `import`. Otherwise, https://github.com/vitejs/vite/issues/16631 would be still broken.

Reply:
Maybe I'm missing something. I would expect the `PackageData` (without the Resolved Cache part) would be the same for both in that case 🤔

Reply:
Good point. Splitting entire `PackageData` seems redundant. Let me iterate a bit more.

Should we go with https://github.com/vitejs/vite/pull/18302 for now? I can follow up separately with your idea.

Reply:
> IIUC making (b) per-environment is not enough. It also needs to be separated whether it's a resolve for `require` or `import`. Otherwise, #16631 would be still broken.

For this, I was thinking we still need to expand `isRequire` as a cache key, which we aren't doing. But, for other flags `conditions, mainFields, ...`, we can assume those are static per environment, so we don't need to include it as cache key like, which was the heavy part of https://github.com/vitejs/vite/pull/18302